pkg/kubelet/cm/topologymanager/topology_manager.go

```diff
@@ -233,43 +251,86 @@ func (m *manager) Admit(attrs *lifecycle.PodAdmitAttributes) lifecycle.PodAdmitR
+	return m.containerAdmit(pod)
+}
 
+func (m *manager) containerAdmit(pod *v1.Pod) lifecycle.PodAdmitResult {
+	for _, container := range append(pod.Spec.InitContainers, pod.Spec.Containers...) {
 		result, admit := m.calculateAffinity(pod, &container)
 		if !admit {
-			return lifecycle.PodAdmitResult{
-				Message: "Resources cannot be allocated with Topology locality",
-				Reason:  "TopologyAffinityError",
-				Admit:   false,
-			}
+			return podAdmitError(topologyAffinityError, nil)
 		}
-
-		klog.Infof("[topologymanager] Topology Affinity for (pod: %v container: %v): %v", pod.UID, container.Name, result)
+		klog.Infof("[topologymanager] Topology Affinity for (pod: %v container: %v): %v", format.Pod(pod), container.Name, result)
 		if m.podTopologyHints[string(pod.UID)] == nil {
 			m.podTopologyHints[string(pod.UID)] = make(map[string]TopologyHint)
 		}
 		m.podTopologyHints[string(pod.UID)][container.Name] = result
-
 		err := m.allocateAlignedResources(pod, &container)
 		if err != nil {
-			return lifecycle.PodAdmitResult{
-				Message: fmt.Sprintf("Allocate failed due to %v, which is unexpected", err),
-				Reason:  "UnexpectedAdmissionError",
-				Admit:   false,
-			}
+			return podAdmitError(unexpectedAdmissionError, err)
 		}
 	}
 
 	return lifecycle.PodAdmitResult{Admit: true}
 }
+
+func (m *manager) podAdmit(pod *v1.Pod) lifecycle.PodAdmitResult {
+	var providersHints []map[string][]TopologyHint
+	for _, provider := range m.hintProviders {
+		// Get the TopologyHints for Pod from a provider.
+		hints := provider.GetPodTopologyHints(pod)
+		providersHints = append(providersHints, hints)
+		klog.Infof("[topologymanager] TopologyHints for pod '%v': %v", format.Pod(pod), hints)
+	}
+	result, admit := m.policy.Merge(providersHints)
+	klog.Infof("[topologymanager] Best TopologyHint for pod : %v", result)
+	if !admit {
+		return podAdmitError(topologyAffinityError, nil)
+	}
+	if m.podTopologyHints[string(pod.UID)] == nil {
+		m.podTopologyHints[string(pod.UID)] = make(map[string]TopologyHint)
+	}
+	for _, container := range append(pod.Spec.InitContainers, pod.Spec.Containers...) {
+		klog.Infof("[topologymanager] Topology Affinity for (pod: %v container: %v): %v", format.Pod(pod), container.Name, result)
+		m.podTopologyHints[string(pod.UID)][container.Name] = result
+		err := m.allocateAlignedResources(pod, &container)
+		if err != nil {
+			return podAdmitError(unexpectedAdmissionError, err)
+		}
+	}
+	return lifecycle.PodAdmitResult{Admit: true}
+}
+
+func podAdmitError(errCode int, err error) lifecycle.PodAdmitResult {
+	switch errCode {
+
+	case topologyAffinityError:
+		return lifecycle.PodAdmitResult{
+			Message: "Resources cannot be allocated with Topology locality",
+			Reason:  "TopologyAffinityError",
+			Admit:   false,
+		}
+
+	default:
+		return lifecycle.PodAdmitResult{
+			Message: fmt.Sprintf("Allocate failed due to %v, which is unexpected", err),
+			Reason:  "UnexpectedAdmissionError",
+			Admit:   false,
+		}
+	}
+}
```

Comment:
Instead of introducing two new variables and then switching on them in this function, I think it probably just makes sense to create two functions:
```
func topologyAffinityError() lifecycle.PodAdmitResult {
	return lifecycle.PodAdmitResult{
		Message: "Resources cannot be allocated with Topology locality",
		Reason:  "TopologyAffinityError",
		Admit:   false,
	}
}

func unexpectedAdmissionError(error err)  lifecycle.PodAdmitResult {
	return lifecycle.PodAdmitResult{
		Message: fmt.Sprintf("Allocate failed due to %v, which is unexpected", err),
		Reason:  "UnexpectedAdmissionError",
		Admit:   false,
	}
}
```

and while we're at it, we might as well create a function for the success case to be consistent:
```
func admitPod() lifecycle.PodAdmitResult {
	return lifecycle.PodAdmitResult{Admit: true}
}
```

Reply:
Done